### PR TITLE
Implement GROUP BY

### PR DESCRIPTION
**Pull request by tuespetre**
# I don't know when to quit.

Imgur

--------

This PR (dependent on #7695 and #7696) implements GROUP BY, this time with ~~only about 15 or 16 changed files~~ <ins>less changed files, although still a good number due to test changes</ins> and a very focused approach to **getting it done without a lot of byproduct.**

### CODE REVIEW

**Comment by smitpatel** [comment]
@tuespetre - Given the 2 dependent PRs are merged in. Can you update/rebase this PR?

**Comment by tuespetre** [comment]
@smitpatel sure can, but heads up, there are some test cases and scenarios I have thought of that still need to be addressed here. I wouldn't bother spending your time reviewing it or anything yet. Sorry ☹️

**Comment by smitpatel** [comment]
Just change the prefix the title with `[WIP]` and update it once you are done working out. Or given there has not been any discussion here. You can close this and open a new PR when work is ready for review.

**Comment by tuespetre** [comment]
OK. Thanks

**Comment by tuespetre** [comment]
@smitpatel I am now going to state that this PR will depend on #7767 and #7771

**Comment by tuespetre** [comment]
:up::date: finished work on tests: split out separate `AsyncGroupByQuery____` and `GroupByQuery___` test classes, and added several new and complex test cases for grouping, plus made them actually work :trollface:

**Comment by smitpatel** [comment]
@tuespetre - The query you posted in PR #7771 , certainly it is a complex query for group join.
What is expected Sql for that?
In general what are the blockers for group by translation in current stack. (not just PR #7771 but other issues you faced which prohibits group by translations)

**Comment by tuespetre** [comment]
@smitpatel here is the query I was basing that on:

https://github.com/tuespetre/EntityFramework/blob/d14681e89b2b2cb95840abc460186167ddcb9a85/test/EFCore.SqlServer.FunctionalTests/GroupByQuerySqlServerTest.cs#L469

I don't see anything else blocking generally useful GROUP BY translations for now. One scenario I have not covered in this PR is aggregate subqueries that include `WhereClause`s (see the `GroupBy_Sum_Where` test in that same file for an example.) I viewed that as an additional enhancement beyond the basic ability to group and aggregate.

**Comment by tuespetre** [comment]
:up::date: ➕  ♻️🅱️

**Comment by tuespetre** [comment]
:up::date: ➕  ♻️🅱️ again

File: src/EFCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs [hunk 1]
```diff
@@ -68,22 +71,59 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
...
+        ///     An Expression corresponding to the translated member.
+        /// </returns>
+        protected override Expression VisitMember(MemberExpression memberExpression)
+        {
+            Check.NotNull(memberExpression, nameof(memberExpression));
+
+            var sqlExpression = TranslateExpression(memberExpression);
+
+            var handledExpression
+                = TryHandleGroupingKeyExpression(
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs]: Remove these (and also methodcall) methods, and call appropriate code from main `Visit`.
It is good to have expression type specific visitors but if they are doing the same task as the main Visit method then it does not serve much value.

**Comment by smitpatel** [comment]
@tuespetre - Please remove any trailing code changes from any other PRs which are not merged. Also for reviewing purpose create 2 different commits. 2nd one containing only test changes (there are 2200 lines of addition just from tests 😄 nice.) And let us know when it is ready for review.

**Comment by smitpatel** [comment]
We are closing this pull request due to lack of response. If you decide to work on the issue again then feel free to reopen the PR and push more commits to it. We would be happy to review it.
Thank you helping us out in improving EF Core.

**Comment by smitpatel** [comment]
@tuespetre - I have taken tests from this PR. Thanks.

**Pull request closed without merging**